crates/matrix-sdk-ui/src/timeline/event_handler.rs

```diff
@@ -655,6 +682,53 @@ impl<'a, 'o> TimelineEventHandler<'a, 'o> {
         }
     }
 
+    /// Handle a beacon_info event and update the `BeaconInfoState`.
+    ///
+    /// TODO: Handle case of `m.beacon_info` event that is identical to the
+    /// first, except with live: false. This stops the live location share.
+    ///
+    /// TODO: How do you search for an existing beacon_info event in the
+    /// timeline by state key?
+    fn handle_beacon_info(
+        &mut self,
+        c: FullStateEventContent<BeaconInfoEventContent>,
+        should_add: bool,
+        user_id: OwnedUserId,
+    ) {
+        let mut beacon_state = BeaconState::new(c, user_id.clone());
+        if let Flow::Remote { event_id, .. } = self.ctx.flow.clone() {
+            // Applying the cache to remote events only because local echoes
+            // don't have an event ID that could be referenced by responses yet.
+            self.meta.beacon_pending_events.apply(&event_id, &mut beacon_state);
+        }
+
+        if should_add {
+            self.add_item(TimelineItemContent::BeaconInfoState(beacon_state));
+        }
+    }
+
+    /// Attach the last known beacon location to the `BeaconInfoState` for a
+    /// specific beacon_info.
+    ///
+    /// A beacon may arrive before its beacon_info event, so we store the beacon
+    /// in a pending list until the beacon_info event arrives.
+    fn handle_beacon(&mut self, c: BeaconEventContent) {
+        // Find the beacon_info event in the timeline and update the beacon state.
+        let found = self.update_timeline_item(&c.relates_to.event_id, |_, event_item| {
+            let beacon_state =
+                as_variant!(event_item.content(), TimelineItemContent::BeaconInfoState)?;
+            Some(event_item.with_content(
+                TimelineItemContent::BeaconInfoState(beacon_state.update_beacon(&c)),
+                None,
+            ))
+        });
+
+        if !found {
+            warn!("Did not find beacon_info event in timeline items. Adding to pending list");
```

Comment:
`warn` should be kept for non-fatal errors.

```suggestion
            debug!("Did not find beacon_info event in timeline items. Adding to pending list");
```